website/addons/dataverse/views/config.py

```diff
@@ -1,152 +1,102 @@
-    result = {
-        'nodeHasAuth': node_settings.has_auth,
-        'userIsOwner': user_is_owner,
-        'userHasAuth': current_user_settings is not None and current_user_settings.has_auth,
-        'urls': serialize_urls(node_settings),
-    }
+    provider = DataverseProvider()
 
-    if node_settings.has_auth:
-        # Add owner's profile info
-        result['urls']['owner'] = web_url_for('profile_view_id',
-            uid=user_settings.owner._primary_key)
-        result.update({
-            'ownerName': user_settings.owner.fullname,
-            'apiToken': user_settings.api_token,
-        })
-        # Add owner's dataverse settings
-        connection = client.connect_from_settings(user_settings)
-        dataverses = client.get_dataverses(connection)
-        result.update({
-            'connected': connection is not None,
-            'dataverses': [
-                {'title': dataverse.title, 'alias': dataverse.alias}
-                for dataverse in dataverses
-            ],
-            'savedDataverse': {
-                'title': node_settings.dataverse,
-                'alias': node_settings.dataverse_alias
-            },
-            'savedDataset': {
-                'title': node_settings.dataset,
-                'doi': node_settings.dataset_doi
-            }
-        })
-    return result
-
-
-def serialize_urls(node_settings):
-    node = node_settings.owner
-    urls = {
-        'create': api_url_for('dataverse_set_user_config'),
-        'set': node.api_url_for('set_dataverse_and_dataset'),
-        'importAuth': node.api_url_for('dataverse_import_user_auth'),
-        'deauthorize': node.api_url_for('deauthorize_dataverse'),
-        'getDatasets': node.api_url_for('dataverse_get_datasets'),
-        'datasetPrefix': 'http://dx.doi.org/',
-        'dataversePrefix': 'http://{0}/dataverse/'.format(HOST),
-        'apiToken': 'https://{0}/account/apitoken'.format(HOST),
-        'settings': web_url_for('user_addons'),
-    }
-    return urls
+    host = request.json.get('host').rstrip('/')
+    api_token = request.json.get('api_token')
+
+    # Verify that credentials are valid
+    client.connect_or_401(host, api_token)
+
+    # Note: `DataverseSerializer` expects display_name to be a URL
+    try:
+        provider.account = ExternalAccount(
+            provider=provider.short_name,
+            provider_name=provider.name,
+            display_name=host,       # no username; show host
+            oauth_key=host,          # hijacked; now host
+            oauth_secret=api_token,  # hijacked; now api_token
+            provider_id=api_token,   # Change to username if Dataverse allows
+        )
+        provider.account.save()
+    except KeyExistsException:
+        # ... or get the old one
+        provider.account = ExternalAccount.find_one(
+            Q('provider', 'eq', provider.short_name) &
+            Q('provider_id', 'eq', api_token)
+        )
+        assert provider.account is not None
```

Comment:
Correct - there is (and should be) no way for `find_one` to return anything other than a subclass of `StoredObject`.

In current Modular-ODM, it can only return an instance of the class upon which it is called. In other branches, and probably in future version of Modular-ODM, it will be possible for it to return subclasses of the class upon which it is called as well - but it should never return `None`.